step-generation/src/utils/safePipetteMovements.ts

```diff
@@ -134,79 +121,83 @@ const getHasOverlappingRectangles = (
-    Math.abs(Math.max(...yCoordinates) - Math.min(...yCoordinates)) <
-    yLengthRect1 + yLengthRect2
+  const oneLeftOfTwo = rectangle1[1].x < rectangle2[0].x
+  const oneRightOfTwo = rectangle1[0].x > rectangle2[1].x
+  const oneUnderTwo = rectangle1[0].y < rectangle2[1].y
+  const oneOverTwo = rectangle1[1].y > rectangle2[0].y
 
-  return overlappingInX && overlappingInY
+  return !(oneLeftOfTwo || oneRightOfTwo || oneUnderTwo || oneOverTwo)
+}
+
+const getModuleHeightFromDeckDefinition = (
+  moduleModel: ModuleModel
+): number => {
+  const deckDef = getDeckDefFromRobotType(FLEX_ROBOT_TYPE)
+  const { addressableAreas } = deckDef.locations
+  const moduleAddressableArea = addressableAreas.find(addressableArea =>
+    addressableArea.id.includes(moduleModel)
+  )
+  return moduleAddressableArea?.offsetFromCutoutFixture[2] ?? 0
 }
 
 //  check the highest Z-point of all items stacked given a deck slot (including modules,
 //  adapters, and modules on adapters)
 const getHighestZInSlot = (
   robotState: RobotState,
   invariantContext: InvariantContext,
-  labwareId: string
+  slotId: string
 ): number => {
   const { modules, labware } = robotState
   const { moduleEntities, labwareEntities } = invariantContext
-  if (modules[labwareId] != null) {
-    const moduleDimensions = getModuleDef2(moduleEntities[labwareId].model)
-      .dimensions
-    return (
-      //  labware + module
-      labwareEntities[labwareId].def.dimensions.zDimension +
-      moduleDimensions.bareOverallHeight +
-      (moduleDimensions.lidHeight ?? 0)
+
+  let totalHeight = 0
+  const moduleInSlot = Object.keys(modules).find(
+    moduleId => modules[moduleId].slot === slotId
+  )
+  // module in slot
+  if (moduleInSlot != null) {
+    totalHeight += getModuleHeightFromDeckDefinition(
+      moduleEntities[moduleInSlot].model
+    )
+    const moduleDirectChildId = Object.keys(labware).find(
+      lwId => labware[lwId].slot === moduleInSlot
     )
-  } else if (labware[labwareId] != null) {
-    const adapterId = labware[labwareId].slot
-    if (labwareEntities[adapterId] != null) {
-      if (modules[adapterId] != null) {
-        const moduleDimensions = getModuleDef2(moduleEntities[adapterId].model)
-          .dimensions
-        return (
-          //  labware + adapter + module
-          labwareEntities[labwareId].def.dimensions.zDimension +
-          labwareEntities[adapterId].def.dimensions.zDimension +
-          moduleDimensions.bareOverallHeight +
-          (moduleDimensions.lidHeight ?? 0)
-        )
-      } else {
-        return (
-          //   labware + adapter
-          labwareEntities[labwareId].def.dimensions.zDimension +
-          labwareEntities[adapterId].def.dimensions.zDimension
-        )
+    if (moduleDirectChildId != null) {
+      const moduleChildHeight =
+        labwareEntities[moduleDirectChildId].def.dimensions.zDimension
+      totalHeight += moduleChildHeight
+
+      // check if adapter and has child
```

Comment:
nit, this is a bit clearer since the same comment is also on line 189
```suggestion
      // check if adapter is on module and has child
```